devito/types/dimension.py

```diff
@@ -395,6 +395,22 @@ def __eq__(self, other):
     __hash__ = Symbol.__hash__
 
 
+class UniqueDimension(BasicDimension):
```

Comment:
I struggle to understand the need for this class

Reply:
`Dimension`s with the same names always hash the same, so when used as dictionary keys, each entry replaces the last, rather than creating new entries. The implicit dimension of the `SubDomainSet` always has the name `'n'`

Reply:
so the actual error is that it should be named uniquely instead ?

Reply:
But then you need a counter again... Although I suppose you could append the `id(self)` to the name when you first create it, before subsequently rebuilding it as `n0`, `n1`, ... Maybe that's the way to go?